Improve hathi_rsync manage command so it can sync and report on all hathi content

rlskoeser opened the pull request:
relates most directly to #428 but also relevant to most of the rsync items aggregated by #598

changes in this PR:
- improves the existing `hathi_rsync` command so it can be run to sync content for all non-suppressed hathi records in PPA
  - fix importer rsync logging so it does not output all source ids when syncing more than a handful of records
  - add an optional output dir for rsync itemized output
- parse rsync itemized output to identify paths that were updated due to a change; convert path to htid and generate a report of all updated htids

I have tested this locally using an rsync from staging server to my local development server and it seems to be working for the scenarios I've tried.

I haven't worked on unit tests related to this change yet, in part because I think this may not be the final form we want - in part because I think this will be one piece in a larger pipeline of updates triggered by content being updated through rsync.

My rsync output parsing does not do any checking for deleted files; that's a different scenario than we're currently concerned about, but we might want to think about that at some point. (... and looks like I made some notes on that on #428)

rlskoeser commented on ppa/archive/import_util.py:
```diff
@@ -247,7 +255,15 @@ def pairtree_paths(self):
         """Use rsync to retrieve data for the volumes to be imported."""
 
-        logger.info("rsyncing pairtree data for %s", ", ".join(self.source_ids))
+        # limit the number of ids included in the log message
+        log_detail = ""
+        rsync_count = len(self.source_ids)
+        if rsync_count <= 12:
+            log_detail = ", ".join(self.source_ids)
+        else:
+            log_detail = "%d volumes" % rsync_count
+
+        logger.info("rsyncing pairtree data for %s", log_detail)
```
previously this was logging _all_ of the source ids, which worked for small-batches of imports or rsync but is obviously unhelpful when it's all the hathitrust records in ppa

laurejt replied:
What's the reasoning for setting the default to 12 specifically?

rlskoeser replied:
@laurejt totally arbitrary! Just a guess at what might be reasonable. I don't think we want more than a few lines of logging worth of ids, it's not particularly meaningful and it's too noisy. (I did not actually check how this looks in the log files). Could change to  10 or 15 if you think that would be a less distracting number. 😁

laurejt commented on ppa/archive/import_util.py:
```diff
@@ -292,7 +309,7 @@ def rsync_data(self):
                 subprocess.run(args=rsync_cmd.split(), check=True)
             except subprocess.CalledProcessError as err:
                 logger.error(
-                    "HathiTrust rsync failed — %s / command: %s"
+                    "HathiTrust rsync failed — %s / command: %s"
```
Did this line actually change? Or, is it just the diff failing.

rlskoeser replied:
there was a weird character here! I removed it but you can't tell what the change is, I guess it was a different kind of whitespace

laurejt commented on ppa/archive/management/commands/hathi_rsync.py:
```diff
@@ -26,22 +30,64 @@ def handle(self, *args, **kwargs):
+        # in the database and not suppressed
+        if htids:
+            digworks = digworks.filter(source_id__in=htids)
+        # NOTE: report here on any skipped ids?
```
It seems worth logging any skipped ids (either because they don't exist or are suppressed)

laurejt commented on ppa/archive/import_util.py:
```diff
@@ -247,7 +255,15 @@ def pairtree_paths(self):
+        # limit the number of ids included in the log message
+        log_detail = ""
+        rsync_count = len(self.source_ids)
+        if rsync_count <= 12:
```
Suggestion: switch to 10 since it's typically interpreted as less specific than 12.
```suggestion
        if rsync_count <= 10:
```

laurejt commented on ppa/archive/management/commands/hathi_rsync.py:
```diff
@@ -26,22 +30,64 @@ def handle(self, *args, **kwargs):
-        )
-        # even if verbosity is zero we want an output file
+        # generate a list of unique source ids from the queryset
+        hathi_ids = digworks.values_list("source_id", flat=True).distinct()
```
Consider renaming the variable `hathi_ids`, it's easy to confuse this with `htids`
```suggestion
        working_htids = digworks.values_list("source_id", flat=True).distinct()
```

laurejt commented on ppa/archive/import_util.py:
```diff
@@ -178,10 +185,11 @@ class HathiImporter(DigitizedWorkImporter):
     )
 
-    def __init__(self, source_ids=None, rsync_output=False):
+    def __init__(self, source_ids=None, rsync_output=False, output_dir=""):
         super().__init__(source_ids)
         # track whether (and how much) rsync output is desired
         self.rsync_output = rsync_output
+        self.output_dir = output_dir
```
For the purposes of initialization, it seems like there should be an additional constraint that if `rsync_output` is set then `output_dir` must also be specified (and exist).

rlskoeser replied:
previously it was putting the file in the current working directory - but you're right, better to be explicit

laurejt commented on ppa/archive/management/commands/hathi_rsync.py:
```diff
@@ -26,22 +30,64 @@ def handle(self, *args, **kwargs):
+        # on which volumes were updated
         htimporter = HathiImporter(
-            source_ids=htids, rsync_output=self.verbosity or True
+            source_ids=hathi_ids, rsync_output=True, output_dir="/tmp"
```
Is `/tmp` a good output directory for this?

rlskoeser replied:
Oh, I meant to look into that and forgot. If and when we set this functionality to run as a cron job, that's not a good place. In staging/production, maybe it could go into the logging directory. It might be good enough for now, while we're assessing.

laurejt commented on ppa/archive/management/commands/hathi_rsync.py:
```diff
@@ -26,22 +30,64 @@ def handle(self, *args, **kwargs):
         logfile = htimporter.rsync_data()
-        self.stdout.write(self.style.SUCCESS("rsync output is in %s" % logfile))
+
+        # read the rsync itemized output to identify records where file
+        # sizes changed
```
Is it possible for a file to change in a way that matters to use without its size reflecting this?

laurejt commented on ppa/archive/management/commands/hathi_rsync.py:
```diff
@@ -26,22 +30,64 @@ def handle(self, *args, **kwargs):
+                    if not filename.endswith(".zip") and not filename.endswith(".xml"):
+                        continue
+                    # reconstruct the hathi id from the filepath
+                    ht_prefix, pairtree_dir = filename.split("/pairtree_root/")
```
```suggestion
                    ht_prefix, pairtree_dir = filename.split("/pairtree_root/", 1)
```

laurejt commented on ppa/archive/management/commands/hathi_rsync.py:
```diff
@@ -26,22 +30,64 @@ def handle(self, *args, **kwargs):
+                    htid = f"{ht_prefix}.{path2id(pairtree_id)}"
+                    updated_ids.add(htid)
+
+        # should this behavior only be when updating all?
+        # if specific htids are specified on the command line, maybe report on them only?
```
It seems like it'd always be useful to know *which* files were updated by rsync, since it's not guaranteed that any are.

laurejt reviewed (commented):
This looks good overall, but I have one concern about whether a change in size encompasses all changes of interest.

rlskoeser commented:
@laurejt thanks for the great feedback. Appreciate the suggestion for the variable name `working_htids`, much clearer!

updated as follows:
- report on skipped htids if specific ids are passed as a parameter
- now using a `tempfile.TemporaryDirectory` instead of `/tmp`
- generating a csv report of updated files that preserves a lot more of the detail from rsync; may not be what we want ultimately, but could be helpful as we're investigating how rsync helps us with the other problems.
- added and tested validation for the output dir parameter for the rsync logging

laurejt commented on ppa/archive/import_util.py:
```diff
@@ -185,10 +188,19 @@ class HathiImporter(DigitizedWorkImporter):
+                raise ValueError("output_dir is required when rsync_output is enabled")
+            elif not os.path.isdir(output_dir):
+                raise ValueError(
+                    "rsync output dir %s is not an existing directory", output_dir
```
```suggestion
                    f"rsync output dir {output_dir} is not an existing directory"
```

laurejt commented on ppa/archive/management/commands/hathi_rsync.py:
```diff
@@ -40,52 +42,88 @@ def handle(self, *args, **kwargs):
+        # bail out if there's nothing to do
+        # (e.g., explicit htids only and none valid)
+        if not working_htids:
+            return
```
It seems worth logging/reporting that there's nothing to synchronize before exiting

rlskoeser replied:
Ah, good catch. In earlier versions it output "synchronizing data for 0 records" but I've changed the order.

laurejt commented on ppa/archive/management/commands/hathi_rsync.py:
```diff
@@ -40,52 +42,88 @@ def handle(self, *args, **kwargs):
+        working_htids = digworks.values_list("source_id", flat=True).distinct()
+
+        # if htids were explicitly specified, report if any are skipped
+        if htids:
+            skipped_htids = set(htids) - set(working_htids)
+            if skipped_htids:
+                self.stdout.write(
+                    self.style.NOTICE(
+                        "Some ids not found in public HathiTrust volumes; skipping %s"
+                        % " ".join(skipped_htids)
+                    )
+                )
```
Perhaps report the total # of volumes skipped? See also comment for line 63.

laurejt commented on ppa/archive/management/commands/hathi_rsync.py:
```diff
@@ -40,52 +42,88 @@ def handle(self, *args, **kwargs):
                     # last element is the filename that was updated
-                    filename = line.rsplit()[-1].strip()
+                    filename = parts[-1]
+                    # itemized info flags preced the filename
```
```suggestion
                    # itemized info flags precede the filename
```

laurejt commented on ppa/archive/management/commands/hathi_rsync.py:
```diff
@@ -40,52 +42,88 @@ def handle(self, *args, **kwargs):
-            outfilename = "ppa_rsync_updated_htids_%s.txt" % datetime.now().strftime(
+        if updated_files:
+            outfilename = "ppa_rsync_changes_%s.csv" % datetime.now().strftime(
                 "%Y%m%d-%H%M%S"
```
To improve readability
```suggestion
            outfilename = "ppa_rsync_changes_{time}.csv".format(
                time=datetime.now().strftime("%Y%m%d-%H%M%S")
```

laurejt commented on ppa/archive/tests/test_import_util.py:
```diff
@@ -269,6 +268,26 @@ def test_rsync_data(self, mocksubprocess):
+    # rsync output requested with non-existent output dir
+    with pytest.raises(ValueError, match="not an existing directory"):
+        # rsync output requested with invalid output dir
+        HathiImporter(rsync_output=True, output_dir="/tmp/foo/bar")
```
So CodeFactor won't complain (there's no reason `tmp` needs to be mentioned)
```suggestion
        HathiImporter(rsync_output=True, output_dir="/foo/bar/baz")
```
